File: LeetCode/!0034. Find First and Last Position of Element in Sorted Array.cpp
/*
0034. Find First and Last Position of Element in Sorted Array

给定一个按照升序排列的整数数组 nums，和一个目标值 target。找出给定目标值在数组中的开始位置和结束位置。
你的算法时间复杂度必须是 O(log n) 级别。

思路：
进行两次二分查找，分别找到左边界和右边界。
找寻左右边界的方法参考https://www.cnblogs.com/kyoner/p/11080078.html
*/

#include <vector>
using namespace std;

class Solution {
public:
    int findLeft(vector<int>& nums, int target) {
        if (nums.size() == 0)
            return -1;

        int left = 0, right = nums.size();
        while (left < right) {
            int mid = (left + right) / 2;
            if (nums[mid] == target) {
                right = mid;
            }
            else if (nums[mid] > target) {
                right = mid;
            }
            else if (nums[mid] < target) {
                left = mid + 1;
            }
        }

        // target比所有值都要大
        if (left >= nums.size())
            return -1;
        if (nums[left] == target)
            return left;
        else
            return -1;
    }

    int findRight(vector<int> &nums, int target) {
        if (nums.size() == 0)
            return -1;

        int left = 0, right = nums.size();
        while (left < right) {
            int mid = (left + right) / 2;
            if (nums[mid] == target) {
                left = mid + 1;
            }
            else if (nums[mid] > target) {
                right = mid;
            }
            else if (nums[mid] < target) {
                left = mid + 1;
            }
        }

        // target比所有值都小
        if (left == 0)
            return -1;
        if (nums[left - 1] == target)
            return left - 1;
        else
            return -1;
    }

    vector<int> searchRange(vector<int>& nums, int target) {
        int left = findLeft(nums, target);
        int right = findRight(nums, target);
        return {left, right};
    }
};

// 二刷，不知道怎么测试用例过不去：[2, 2], 3
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> res({-1, -1});
        if (nums.size() == 0)
            return res;

        int left = 0, right = nums.size();
        int mid;
        // 找左边界
        while (left < right) {
            mid = left + (right - left) / 2;
            if (nums[mid] == target) {
                right = mid;
            }
            else if (nums[mid] > target) {
                right = mid;
            }
            else if (nums[mid] < target) {
                left = mid + 1;
            }
        }
        if (nums[right] == target) {
            res[0] = right;
        }

        // 找右边界
        left = 0; right = nums.size();
        while (left < right) {
            mid = left + (right - left) / 2;
            if (nums[mid] == target) {
                left = mid + 1;
            }
            else if (nums[mid] > target) {
                right = mid;
            }
            else if (nums[mid] < target) {
                left = mid + 1;
            }
        }
        if (left - 1 >= 0 && nums[left - 1] == target) {
            res[1] = left - 1;
        }
        
        return res;
    }
};

File: LeetCode/!0034. Find First and Last Position of Element in Sorted Array.py
'''
0034. Find First and Last Position of Element in Sorted Array

给定一个按照升序排列的整数数组 nums，和一个目标值 target。
找出给定目标值在数组中的开始位置和结束位置。

思路：
二分法典例！！
进行两次二分查找，分别找到起点和终点。
找到起点和终点的思路参考：
https://www.cnblogs.com/kyoner/p/11080078.html

'''

from typing import List


class Solution:
    def binary_search(self, nums, target, mode):
        if len(nums) == 0:
            return -1
        
        left = 0
        right = len(nums)

        if mode == 'left':
            while left < right:
                mid = int(right - (right - left) / 2)
                if nums[mid] == target:
                    right = mid
                elif nums[mid] < target:
                    left = mid + 1
                elif nums[mid] > target:
                    right = mid
            if left >= len(nums):
                return -1
            if nums[left] == target:
                return left
            else:
                return -1

        elif mode == 'right':
            while left < right:
                mid = int(right - (right - left) / 2)
                if nums[mid] == target:
                    left = mid + 1
                elif nums[mid] < target:
                    left = mid + 1
                elif nums[mid] > target:
                    right = mid
            if right < 1:
                return -1
            elif nums[right - 1] == target:
                return right - 1
            else:
                return -1
    
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        left = self.binary_search(nums, target, 'left')
        right = self.binary_search(nums, target, 'right')
        return [left, right]


sol = Solution()
res = sol.searchRange([1], 1)
print(res)
